File: rcc/services/hardware_service.py
from math import floor
import os
from threading import Lock, Thread
import time

from psutil import Process
import pyudev

from framework.logger import Logger
from rcc.communications.client.cmd.linux.lsusb_client import LS_USB_CLIENT
from rcc.communications.client.dbus.asus.armoury.intel.pl1_spl_client import PL1_SPL_CLIENT
from rcc.communications.client.dbus.asus.armoury.intel.pl2_sppt_client import PL2_SPPT_CLIENT
from rcc.communications.client.dbus.asus.armoury.nvidia.nv_boost_client import NV_BOOST_CLIENT
from rcc.communications.client.dbus.asus.armoury.nvidia.nv_temp_client import NV_TEMP_CLIENT
from rcc.communications.client.dbus.asus.armoury.panel_overdrive_client import PANEL_OVERDRIVE_CLIENT
from rcc.communications.client.dbus.asus.core.platform_client import PLATFORM_CLIENT
from rcc.communications.client.dbus.linux.power_management_keyboard_brightness_control import (
    KEYBOARD_BRIGHTNESS_CONTROL,
)
from rcc.communications.client.dbus.linux.switcheroo_client import SWITCHEROO_CLIENT
from rcc.communications.client.dbus.linux.upower_client import UPOWER_CLIENT
from rcc.communications.client.file.linux.boost_client import BOOST_CLIENT
from rcc.communications.client.file.linux.cpuinfo_client import CPU_INFO_CLIENT
from rcc.communications.client.file.linux.scheduler_client import SSD_SCHEDULER_CLIENT
from rcc.communications.client.tcp.openrgb.openrgb_client import OPEN_RGB_CLIENT
from rcc.gui.notifier import NOTIFIER
from rcc.models.battery_threshold import BatteryThreshold
from rcc.models.cpu_brand import CpuBrand
from rcc.models.gpu_brand import GpuBrand
from rcc.models.performance_profile import PerformanceProfile
from rcc.models.ssd_scheduler import SsdScheduler
from rcc.models.usb_identifier import UsbIdentifier
from rcc.utils.beans import EVENT_BUS, TRANSLATOR
from rcc.utils.events import (
    HARDWARE_SERVICE_BATTERY_THRESHOLD_CHANGED,
    HARDWARE_SERVICE_ON_BATTERY,
    HARDWARE_SERVICE_ON_USB_CHANGED,
    STEAM_SERVICE_GAME_EVENT,
)
from rcc.utils.shell import SHELL


class HardwareService:
    """Hardware service"""

    CPU_PRIORITY = -15
    IO_PRIORITY = floor((CPU_PRIORITY + 20) / 5)
    IO_CLASS = 2

    def __init__(self):  # pylint: disable=too-many-branches
        self._logger = Logger()
        self._logger.info("Initializing HardwareService")
        self._logger.add_tab()

        self._logger.info("Detecting CPU")
        self._logger.add_tab()
        self.__cpu = None
        output = CPU_INFO_CLIENT.read(5)
        if "GenuineIntel" in output:
            self.__cpu = CpuBrand.INTEL
            line = output.splitlines()[-1]
            line = line[line.index(":") + 1 :].strip()
            self._logger.info(line)
        self._logger.add_tab()

        if self.__cpu == CpuBrand.INTEL:
            if PL1_SPL_CLIENT.available:
                self._logger.info("TDP control available")

            if BOOST_CLIENT.available:
                self._logger.info("Boost control available")

        self._logger.rem_tab()
        self._logger.rem_tab()

        self.__gpus = {}
        if SWITCHEROO_CLIENT.available:  # pylint: disable=too-many-nested-blocks
            self._logger.info("Detecting GPU")
            self._logger.add_tab()
            gpus = list(SWITCHEROO_CLIENT.gpus)
            for gpu in sorted(gpus, key=lambda x: (x["Name"])):
                self._logger.info(f"{gpu["Name"]}")
                if gpu["Discrete"]:
                    brand = GpuBrand(gpu["Name"].split(" ")[0].lower())
                    env = None
                    if len(gpu["Environment"]) > 1:
                        env = ""
                        for i in range(0, len(gpu["Environment"]), 2):
                            env += gpu["Environment"][i] + "=" + gpu["Environment"][i + 1] + " "
                        env = env.strip()

                    self.__gpus[brand] = env
                    self._logger.add_tab()

                    if brand == GpuBrand.NVIDIA:
                        if NV_BOOST_CLIENT.available:
                            self._logger.info("Dynamic boost control available")
                        if NV_TEMP_CLIENT.available:
                            self._logger.info("Throttle temperature control available")

                    self._logger.rem_tab()

        self._logger.rem_tab()

        if SSD_SCHEDULER_CLIENT.available:
            self._logger.info("Getting available SSD schedulers")
            self._logger.add_tab()
            self.__available_ssd_sched = SSD_SCHEDULER_CLIENT.get_schedulers()
            for sched in self.__available_ssd_sched:
                self._logger.info(f"{sched.name} - {sched.value}")
            self._logger.rem_tab()

        if UPOWER_CLIENT.available:
            self.__on_bat = UPOWER_CLIENT.on_battery
            UPOWER_CLIENT.on_battery_change(self._on_ac_battery_change)

        if PLATFORM_CLIENT.available:
            self.__battery_charge_limit = PLATFORM_CLIENT.charge_control_end_threshold

        self.__running_games = 0

        self._connected_usb: list[UsbIdentifier] = []
        self._usb_mutex = Lock()
        thread = Thread(name="UsbChecker", target=self.__monitor_for_usb)
        thread.start()

        if KEYBOARD_BRIGHTNESS_CONTROL.available:
            KEYBOARD_BRIGHTNESS_CONTROL.on_brightness_change(self._on_kb_brightness_change)

        EVENT_BUS.on(STEAM_SERVICE_GAME_EVENT, self.__on_game_event)

        self._logger.rem_tab()

    def _on_kb_brightness_change(self, value: int):
        if value == 0 and KEYBOARD_BRIGHTNESS_CONTROL.available:
            KEYBOARD_BRIGHTNESS_CONTROL.keyboard_brightness = 2

    def get_gpu_selector_env(self, gpu: GpuBrand):
        """Get ENV configuration for GPU selection"""
        env = f"RCC_GPU={gpu.value} "

        icds = [
            icd
            for icd in [
                f"/usr/share/vulkan/icd.d/{gpu.value}_icd.i686.json",
                f"/usr/share/vulkan/icd.d/{gpu.value}_icd.x86_64.json",
            ]
            if os.path.exists(icd)
        ]
        if len(icds) > 0:
            env += f"VK_ICD_FILENAMES={":".join(icds)} "

        ocds = [ocd for ocd in [f"/etc/OpenCL/vendors/{gpu.value}.icd"] if os.path.exists(ocd)]
        if len(ocds) > 0:
            env += f"OCL_ICD_FILENAMES={":".join(ocds)} "

        if gpu in self.__gpus and self.__gpus[gpu] is not None:
            env += self.__gpus[gpu]

        return env.strip()

    @property
    def gpus(self) -> list[GpuBrand]:
        """GPU brand"""
        return self.__gpus.keys()

    @property
    def cpu(self) -> CpuBrand | None:
        """CPU brand"""
        return self.__cpu

    def _on_ac_battery_change(self, on_battery: bool, muted=False):
        self.__on_bat = on_battery
        if self.__running_games == 0:
            self.set_panel_overdrive(self.__running_games > 0 and not self.__on_bat)

        if self.__running_games == 0:
            if not muted:
                self._logger.info(
                    f"AC {"un" if on_battery else ""}plugged, battery {"dis" if not on_battery else ""}engaged"
                )
                self._logger.add_tab()
            EVENT_BUS.emit_sequencial(HARDWARE_SERVICE_ON_BATTERY, on_battery)
            if not muted:
                self._logger.rem_tab()

    def __on_game_event(self, count):
        self.__running_games = count

    @property
    def on_battery(self):
        """On battery flag"""
        return self.__on_bat

    @property
    def battery_charge_limit(self) -> BatteryThreshold:
        """Get current battery charge limit"""
        return self.__battery_charge_limit

    def set_battery_threshold(self, value: BatteryThreshold) -> None:
        """Set battery charge threshold"""
        if PLATFORM_CLIENT.available:
            if value != self.__battery_charge_limit:
                PLATFORM_CLIENT.charge_control_end_threshold = value
                self.__battery_charge_limit = value
                EVENT_BUS.emit(HARDWARE_SERVICE_BATTERY_THRESHOLD_CHANGED, value)
                NOTIFIER.show_toast(TRANSLATOR.translate("applied.battery.threshold", {"value": value.value}))

    def set_ssd_scheduler(self, scheduler: SsdScheduler):
        """Set SSD queue scheduler"""
        if SSD_SCHEDULER_CLIENT.available and scheduler in self.__available_ssd_sched:
            self._logger.info(f"SSD scheduler: {scheduler.name}")
            SSD_SCHEDULER_CLIENT.set_scheduler(scheduler)

    def set_boost_status(self, enabled: bool):
        """Enable/disable cpu boost"""
        if BOOST_CLIENT.available:
            self._logger.info(f"CPU boost: {"ENABLED" if enabled else "DISABLED"}")
            BOOST_CLIENT.boost = enabled

    def __monitor_for_usb(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
        """Monitor for usb devices changes"""
        monitor = pyudev.Monitor.from_netlink(pyudev.Context())
        monitor.filter_by("usb")

        lsusb_output = LS_USB_CLIENT.get_usb_dev()
        current_usb = []
        for line in lsusb_output.split("\n"):
            columns = line.strip().split(" ")

            id_vendor, id_product = columns[5].split(":")
            name = " ".join(columns[6:])

            usb_dev = UsbIdentifier(id_vendor, id_product, name)

            for cd in OPEN_RGB_CLIENT.compatible_devices:
                if cd.id_vendor == usb_dev.id_vendor and cd.id_product == usb_dev.id_product:
                    self._connected_usb.append(cd)

        for action, _ in monitor:  # pylint: disable=too-many-nested-blocks
            if action in ["add", "remove"]:
                self._usb_mutex.acquire(True)  # pylint: disable=consider-using-with
                try:
                    lsusb_output = LS_USB_CLIENT.get_usb_dev()

                    current_usb = []
                    for line in lsusb_output.split("\n"):
                        columns = line.strip().split(" ")

                        id_vendor, id_product = columns[5].split(":")
                        name = " ".join(columns[6:])

                        usb_dev = UsbIdentifier(id_vendor, id_product, name)

                        if any(
                            cd.id_vendor == usb_dev.id_vendor and cd.id_product == usb_dev.id_product
                            for cd in OPEN_RGB_CLIENT.compatible_devices
                        ):
                            current_usb.append(usb_dev)

                    added = []
                    for dev1 in current_usb:
                        found = False
                        for dev2 in self._connected_usb:
                            if not found and dev1.id_vendor == dev2.id_vendor and dev1.id_product == dev2.id_product:
                                found = True
                        if not found:
                            added.append(dev1)

                    removed = []
                    for dev1 in self._connected_usb:
                        found = False
                        for dev2 in current_usb:
                            if not found and dev1.id_vendor == dev2.id_vendor and dev1.id_product == dev2.id_product:
                                found = True
                        if not found:
                            removed.append(dev1)

                    if len(removed) > 0:
                        self._logger.info("Removed compatible device(s):")
                        self._logger.add_tab()
                        for item in removed:
                            self._logger.info(OPEN_RGB_CLIENT.get_device_name(item.id_vendor, item.id_product))
                            OPEN_RGB_CLIENT.disable_device(item.name)
                        self._logger.rem_tab()

                    if len(added) > 0:
                        self._logger.info("Connected compatible device(s):")
                        self._logger.add_tab()
                        for item in added:
                            self._logger.info(OPEN_RGB_CLIENT.get_device_name(item.id_vendor, item.id_product))
                        self._logger.rem_tab()

                        self._logger.add_tab()
                        EVENT_BUS.emit_sequencial(HARDWARE_SERVICE_ON_USB_CHANGED)
                        self._logger.rem_tab()
                        self._usb_mutex.release()
                    else:
                        self._usb_mutex.release()

                    self._connected_usb = current_usb
                except Exception:
                    self._usb_mutex.release()

    def set_panel_overdrive(self, enabled):
        """Enable or disable panel overdrive"""
        if PANEL_OVERDRIVE_CLIENT.available:
            self._logger.info(f"Setting panel overdrive {"disabled" if not enabled else "enabled"}")
            PANEL_OVERDRIVE_CLIENT.current_value = 1 if enabled else 0

    def apply_proccess_optimizations(self, pid: int, first_run=True):
        """Apply optimizations to the process tree"""
        if first_run:
            self._logger.info(
                f"Setting CPU priority to {self.CPU_PRIORITY} and " + f"IO priority to {self.IO_PRIORITY}"
            )

        pending = [pid]
        processed = []

        while True:
            new_pids_found = False

            while pending:
                pid = pending.pop(0)
                if pid not in processed:
                    try:
                        SHELL.run_command(
                            f"renice -n {self.CPU_PRIORITY} -p {pid} && "
                            + f"ionice -c {self.IO_CLASS} -n {self.IO_PRIORITY} -p {pid}",
                            sudo=True,
                            check=True,
                        )
                    except Exception as e:
                        self._logger.debug(f"Could not apply optimizations for process {pid}: {e}")
                    processed.append(pid)

                try:
                    for child in Process(pid).children():
                        if child.pid not in pending and child.pid not in processed:
                            pending.append(child.pid)
                            new_pids_found = True
                except Exception:
                    pass

            if not new_pids_found:
                break

            pending = processed
            processed = []

            time.sleep(0.05)

        if first_run:
            self._logger.info(f"Optimized {len(processed)} processes")

        return len(processed)

    def set_cpu_tdp(self, profile: PerformanceProfile):
        """Set CPU TDP configuration"""
        if HARDWARE_SERVICE.cpu == CpuBrand.INTEL:
            pl1 = profile.ac_intel_pl1_spl
            pl2 = profile.ac_intel_pl2_sppt
            if pl1 is not None:
                time.sleep(0.05)
                self._logger.info(f"{self.cpu.capitalize()} CPU")

                if UPOWER_CLIENT.available and UPOWER_CLIENT.on_battery:
                    pl1 = profile.battery_intel_pl1_spl
                    pl2 = profile.battery_intel_pl2_sppt

                self._logger.info(f"  PL1: {pl1}W")
                PL1_SPL_CLIENT.current_value = PL1_SPL_CLIENT.default_value
                time.sleep(0.05)
                PL1_SPL_CLIENT.current_value = pl1

                if pl2 is not None:
                    self._logger.info(f"  PL2: {pl2}W")
                    PL2_SPPT_CLIENT.current_value = PL2_SPPT_CLIENT.default_value
                    time.sleep(0.05)
                    PL2_SPPT_CLIENT.current_value = pl2

    def set_gpu_tgp(self, profile: PerformanceProfile):
        """Set GPU TGP parameters"""
        if GpuBrand.NVIDIA in self.__gpus:
            nv = profile.ac_nv_boost
            nt = profile.ac_nv_temp
            if nv is not None or nt is not None:
                self._logger.info("Nvidia GPU")
                if UPOWER_CLIENT.available and UPOWER_CLIENT.on_battery:
                    nv = profile.battery_nv_boost
                    nt = profile.battery_nv_temp

                if nv is not None:
                    self._logger.info(f"  Dynamic Boost: {nv}W")
                    NV_BOOST_CLIENT.current_value = nv

                if nt is not None:
                    self._logger.info(f"  Throttle temp: {nt}ºC")
                    NV_TEMP_CLIENT.current_value = nt

    @property
    def tdp_available(self):
        """Flag for TDP support"""
        return (GpuBrand.NVIDIA in self.__gpus and (NV_BOOST_CLIENT.available or NV_TEMP_CLIENT.available)) or (
            HARDWARE_SERVICE.cpu == CpuBrand.INTEL and PL1_SPL_CLIENT.available
        )


HARDWARE_SERVICE = HardwareService()

File: RccCore/src/services/hardware_service.cpp
#include "../../include/services/hardware_service.hpp"

#include "../../include/clients/dbus/asus/armoury/intel/pl1_spd_client.hpp"
#include "../../include/clients/dbus/asus/armoury/nvidia/nv_boost_client.hpp"
#include "../../include/clients/dbus/asus/armoury/nvidia/nv_temp_client.hpp"
#include "../../include/clients/dbus/asus/armoury/other/panel_overdrive_client.hpp"
#include "../../include/clients/dbus/asus/core/platform_client.hpp"
#include "../../include/clients/dbus/linux/notifications_client.hpp"
#include "../../include/clients/dbus/linux/power_management_kb_brightness.hpp"
#include "../../include/clients/dbus/linux/upower_client.hpp"
#include "../../include/clients/file/boost_control_client.hpp"
#include "../../include/clients/file/cpuinfo_client.hpp"
#include "../../include/clients/file/ssd_scheduler_client.hpp"
#include "../../include/clients/shell/lsusb_client.hpp"
#include "../../include/clients/shell/switcherooctl_client.hpp"
#include "../../include/clients/tcp/open_rgb/open_rgb_client.hpp"
#include "../../include/gui/toaster.hpp"
#include "../../include/models/hardware/battery_charge_threshold.hpp"
#include "../../include/models/hardware/cpu_brand.hpp"
#include "../../include/models/hardware/gpu_brand.hpp"
#include "../../include/services/open_rgb_service.hpp"
#include "../../include/translator/translator.hpp"
#include "RccCommons.hpp"

HardwareService::HardwareService() {
	logger.info("Initializing HardwareService");
	logger.add_tab();

	logger.info("Detecting CPU");
	logger.add_tab();
	std::string cpuinfo_out = CPUInfoClient::getInstance().read(5);
	if (StringUtils::isSubstring("GenuineIntel", cpuinfo_out)) {
		cpu		   = CpuBrand::Enum::INTEL;
		auto lines = StringUtils::splitLines(cpuinfo_out);
		auto line  = lines[lines.size() - 1];
		auto pos   = line.find(":");
		if (pos != std::string::npos) {
			line = line.substr(pos + 2);
			logger.info(line);
		}
	}

	logger.add_tab();
	if (cpu == CpuBrand::Enum::INTEL) {
		if (Pl1SpdClient::getInstance().available()) {
			logger.info("TDP control available");
		}
		if (BoostControlClient::getInstance().available()) {
			logger.info("Boost control available");
		}
	}
	logger.rem_tab();
	logger.rem_tab();

	logger.info("Detecting GPUs");
	logger.add_tab();
	auto detected_gpus = SwitcherooCtlClient::getInstance().getGpus();
	for (auto gpu : detected_gpus) {
		logger.info(gpu.name);
		if (!gpu.default_flag) {
			auto brand = static_cast<GpuBrand>(GpuBrand::fromString(StringUtils::toLowerCase(StringUtils::split(gpu.name, ' ')[0])));
			std::string env;
			if (!gpu.environment.empty()) {
				for (auto gpu_env : gpu.environment)
					env = env + gpu_env + " ";
			}

			if (FileUtils::exists(Constants::LIB_VK_DIR))
				FileUtils::remove(Constants::LIB_VK_DIR);
			FileUtils::mkdirs(Constants::LIB_VK_DIR);

			if (FileUtils::exists(Constants::LIB_OCL_DIR))
				FileUtils::remove(Constants::LIB_OCL_DIR);
			FileUtils::mkdirs(Constants::LIB_OCL_DIR);

			std::vector<std::string> vkIcd;
			std::vector<std::string> vkIcdVariants = {brand.toString() + "_icd.json", brand.toString() + "_icd.i686.json",
													  brand.toString() + "_icd.x86_64.json"};
			for (auto var : vkIcdVariants) {
				if (FileUtils::exists(Constants::USR_SHARE_VK_DIR + var)) {
					FileUtils::copy(Constants::USR_SHARE_VK_DIR + var, Constants::LIB_VK_DIR + var);
					vkIcd.emplace_back(Constants::LIB_VK_DIR + var);
				}
			}
			if (!vkIcd.empty()) {
				std::ostringstream oss;

				for (size_t i = 0; i < vkIcd.size(); ++i) {
					if (i > 0)
						oss << ":";
					oss << vkIcd[i];
				}

				env = env + "VK_ICD_FILENAMES=" + oss.str() + " ";
			}

			if (FileUtils::exists(Constants::USR_SHARE_OCL_DIR + brand.toString() + ".icd")) {
				FileUtils::copy(Constants::USR_SHARE_OCL_DIR + brand.toString() + ".icd", Constants::LIB_OCL_DIR + brand.toString() + ".icd");
				env = env + "OCL_ICD_FILENAMES=" + Constants::LIB_OCL_DIR + brand.toString() + ".icd" + " ";
			}
			env					   = StringUtils::trim(env);
			gpus[brand.toString()] = env;

			logger.add_tab();
			if (brand == GpuBrand::Enum::NVIDIA) {
				if (NvBoostClient::getInstance().available()) {
					logger.info("Dynamic boost control available");
				}
				if (NvTempClient::getInstance().available()) {
					logger.info("Throttle temperature control available");
				}
			}

			logger.rem_tab();
		}
	}
	logger.rem_tab();

	if (SsdSchedulerClient::getInstance().available()) {
		logger.info("Getting available SSD schedulers");
		logger.add_tab();
		ssd_schedulers = SsdSchedulerClient::getInstance().get_schedulers();
		for (auto sched : ssd_schedulers)
			logger.info(sched.toString());
		logger.rem_tab();
	}

	if (PlatformClient::getInstance().available()) {
		logger.info("Getting battery charge limit");
		logger.add_tab();
		charge_limit = PlatformClient::getInstance().getBatteryLimit();
		logger.info(std::to_string(charge_limit.toInt()) + "%");
		logger.rem_tab();
	}

	setupDeviceLoop();

	runningGames = 0;

	if (UPowerClient::getInstance().available()) {
		onBattery = UPowerClient::getInstance().isOnBattery();
		UPowerClient::getInstance().onBatteryChange([this](CallbackParam data) { this->onBatteryEvent(std::any_cast<bool>(data[0])); });
	}

	if (PMKeyboardBrightness::getInstance().available()) {
		PMKeyboardBrightness::getInstance().onBrightnessChange([this]() {
			if (PMKeyboardBrightness::getInstance().getKeyboardBrightness() == 0) {
				PMKeyboardBrightness::getInstance().setKeyboardBrightnessSilent(2);
			}
		});
	}

	logger.rem_tab();
}

void HardwareService::setupDeviceLoop() {
	auto& udevClient = LsUsbClient::getInstance();
	connectedDevices = std::get<0>(
		udevClient.compare_connected_devs({}, [](const UsbIdentifier& dev) { return !OpenRgbService::getInstance().getDeviceName(dev).empty(); }));
	EventBus::getInstance().on_without_data(Events::UDEV_CLIENT_DEVICE_EVENT, [this, &udevClient]() {
		auto [current, added, removed] = udevClient.compare_connected_devs(
			connectedDevices, [](const UsbIdentifier& dev) { return !OpenRgbService::getInstance().getDeviceName(dev).empty(); });

		if (added.size() > 0) {
			logger.info("Added compatible device(s):");
			logger.add_tab();
			for (auto dev : added) {
				logger.info(OpenRgbService::getInstance().getDeviceName(dev));
			}
			logger.rem_tab();
		}

		if (removed.size() > 0) {
			logger.info("Removed compatible device(s):");
			logger.add_tab();
			for (auto dev : removed) {
				logger.info(OpenRgbService::getInstance().getDeviceName(dev));
			}
			logger.rem_tab();
		}

		if (removed.size() > 0 && added.size() == 0) {
			for (auto dev : removed) {
				OpenRgbService::getInstance().disableDevice(dev);
			}
		} else if (removed.size() > 0 || added.size() > 0) {
			EventBus::getInstance().emit_event(Events::HARDWARE_SERVICE_USB_ADDED_REMOVED);
		}

		connectedDevices = current;
	});
}

BatteryThreshold HardwareService::getChargeThreshold() {
	return charge_limit;
}

void HardwareService::setChargeThreshold(const BatteryThreshold& threshold) {
	std::lock_guard<std::mutex> lock(actionMutex);
	if (charge_limit != threshold) {
		logger.info("Setting charge limit to " + std::to_string(threshold.toInt()) + "%");
		auto t0 = std::chrono::high_resolution_clock::now();
		PlatformClient::getInstance().setBatteryLimit(threshold);
		auto t1 = std::chrono::high_resolution_clock::now();

		charge_limit = threshold;
		logger.info("Charge limit setted after " + std::to_string(std::chrono::duration_cast<std::chrono::milliseconds>(t1 - t0).count()) + " ms");

		std::unordered_map<std::string, std::any> replacements = {{"value", threshold.toInt()}};
		Toaster::getInstance().showToast(Translator::getInstance().translate("applied.battery.threshold", replacements));
		EventBus::getInstance().emit_event(Events::HARDWARE_SERVICE_THRESHOLD_CHANGED, {threshold});
	}
}

void HardwareService::onBatteryEvent(const bool& onBat, const bool& muted) {
	onBattery = onBat;

	setPanelOverdrive(runningGames > 0 && !onBattery);

	if (runningGames == 0) {
		if (!muted) {
			std::string t1 = onBattery ? "un" : "";
			std::string t2 = !onBattery ? "dis" : "";
			logger.info("AC " + t1 + "plugged, battery " + t2 + "engaged");
			logger.add_tab();
		}
		EventBus::getInstance().emit_event(Events::HARDWARE_SERVICE_ON_BATTERY);
		if (!muted) {
			logger.rem_tab();
		}
	}
}

void HardwareService::setPanelOverdrive(const bool& enable) {
	if (PanelOverdriveClient::getInstance().available()) {
		logger.info("Setting panel overdrive to {}", enable);
		PanelOverdriveClient::getInstance().setCurrentValue(enable);
	}
}
void HardwareService::renice(const pid_t& pid) {
	Shell::getInstance().run_elevated_command(
		fmt::format("renice -n {} -p {} && ionice -c {} -n {} -p {}", CPU_PRIORITY, pid, IO_CLASS, IO_PRIORITY, pid));
}

std::map<std::string, std::string> HardwareService::getGpuSelectorEnv(const std::string& gpu) {
	std::map<std::string, std::string> env;

	auto it = gpus.find(gpu);
	if (it != gpus.end()) {
		auto gpuEnv = it->second;

		std::istringstream ss(gpuEnv);
		std::string token;
		std::map<std::string, std::string> env_vars;

		while (ss >> token) {
			auto pos = token.find('=');
			if (pos != std::string::npos) {
				std::string key	  = token.substr(0, pos);
				std::string value = token.substr(pos + 1);
				env_vars[key]	  = value;
			}
		}

		for (const auto& [key, val] : env_vars) {
			env[key] = val;
		}
	}

	return env;
}